src/methods/create.ts

```diff
@@ -1,86 +1,146 @@
+  Components,
+  IDocumentsStore,
+  IIndex,
+  Lyra,
+  OpaqueDocumentStore,
+  OpaqueIndex,
+  Schema,
+  SimpleComponents,
+  SimpleOrArrayCallbackComponents,
+} from "../types.js";
+import { createDocumentsStore } from "../components/documents-store.js";
 
-  buildIndex(instance, properties.schema);
-  return instance;
+interface CreateArguments<S extends Schema, I extends OpaqueIndex, D extends OpaqueDocumentStore> {
+  schema: Schema;
+  defaultLanguage?: string;
+  components?: Components<S, I, D>;
 }
 
-function buildIndex<S extends PropertiesSchema>(lyra: Lyra<S>, schema: S, prefix = "") {
-  for (const prop of Object.keys(schema)) {
-    const propType = typeof prop;
-    const isNested = typeof schema[prop] === "object";
-
-    if (propType !== "string") throw new Error(ERRORS.INVALID_SCHEMA_TYPE(propType));
+function validateComponents<S extends Schema, I extends OpaqueIndex, D extends OpaqueDocumentStore>(
+  components: Components<S, I, D>,
+) {
+  const defaultComponents = getDefaultComponents();
 
-    const propName = `${prefix}${prop}`;
+  for (const rawKey of SIMPLE_COMPONENTS) {
+    const key = rawKey as keyof SimpleComponents;
 
-    if (isNested) {
-      buildIndex(lyra, schema[prop] as S, `${propName}.`);
-    } else {
-      if (schema[prop] === "string") {
-        lyra.index[propName] = createNode();
-        lyra.avgFieldLength[propName] = 0;
-        continue;
-      }
-      
-      if (schema[prop] === "number") {
-        lyra.index[propName] = createAVLNode<number, string[]>(0, []);
-        continue;
+    if (components[key]) {
+      if (typeof components[key] !== "function") {
+        throw createError("COMPONENT_MUST_BE_FUNCTION", key);
       }
+    } else {
+      // @ts-expect-error TSC is unable to resolve this
+      components[key] = defaultComponents[key];
+    }
+  }
+
+  for (const rawKey of SIMPLE_OR_ARRAY_COMPONENTS) {
+    const key = rawKey as keyof ArrayCallbackComponents<S, I, D>;
 
-      if (schema[prop] === "boolean") {
-        lyra.index[propName] = { 'true': [], 'false': [] };
-        continue;
+    if (!components[key]) {
+      components[key] = [];
+    } else if (!Array.isArray(components[key])) {
+      // @ts-expect-error TSC is unable to resolve this
+      components[key] = [components[key]];
+    }
+
+    for (const fn of components[key] as unknown as SimpleOrArrayCallbackComponents<S, I, D>[]) {
+      if (typeof fn !== "function") {
+        throw createError("COMPONENT_MUST_BE_FUNCTION_OR_ARRAY_FUNCTIONS", key);
       }
     }
   }
+
+  for (const rawKey of Object.keys(components)) {
+    if (
+      !COMPLEX_COMPONENTS.includes(rawKey) &&
+      !SIMPLE_COMPONENTS.includes(rawKey) &&
+      !SIMPLE_OR_ARRAY_COMPONENTS.includes(rawKey)
+    ) {
+      throw createError("UNSUPPORTED_COMPONENT", rawKey);
+    }
+  }
+}
+
+export async function create<S extends Schema, I extends OpaqueIndex, D extends OpaqueDocumentStore>({
+  schema,
+  defaultLanguage,
```

Comment:
I thought we removed `defaultLanguage` in favor of `language`?

Reply:
Yes we did. At the beginning I was try to maintain compatibility but then I gave up. I'll rename this.

Reply:
Done. Will be there in next commit.